### PR TITLE
feat: Support graceful shutdown

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

Part of https://github.com/stackabletech/issues/issues/357
Closes https://github.com/stackabletech/hbase-operator/issues/400

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] CRD changes approved
- [x] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [x] (Integration-)Test cases added
- [x] Documentation added or updated
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

File: docs/modules/hbase/pages/usage-guide/operations/graceful-shutdown.adoc [hunk 1]
```diff
@@ -1,7 +1,41 @@
 = Graceful shutdown
 
-Graceful shutdown of HBase nodes is either not supported by the product itself
-or we have not implemented it yet.
+You can configure the graceful shutdown as described in xref:concepts:operations/pod_placement.adoc[].
```
Comment (sbernauer) [inline docs/modules/hbase/pages/usage-guide/operations/graceful-shutdown.adoc:3]:
```suggestion
You can configure the graceful shutdown as described in xref:concepts:operations/graceful_shutdown.adoc[].
```

File: docs/modules/hbase/pages/usage-guide/operations/graceful-shutdown.adoc [hunk 2]
```diff
@@ -1,7 +1,41 @@
...
+
+This is equivalent to executing the `bin/hbase-daemon.sh stop master` command, which will internally also just `kill <master-pid>` (https://github.com/apache/hbase/blob/8382f55b15be6ae190f8d202a5e6a40af177ec76/bin/hbase-daemon.sh#L338[code]), wait for a configurable timeout (defaults to 20 minutes) after which `kill -9 <master-pid>` is used to `SIGKILL` the master (https://github.com/apache/hbase/blob/8382f55b15be6ae190f8d202a5e6a40af177ec76/bin/hbase-common.sh#L20-L41[code]).
+
+However, there is no acknowledge message in the log indicating a graceful shutdown.
+
+== RegionServers
+
+As a default, RegionServers have `60 minutes` to terminate gracefully.
+
+They go through the same mechanism as documented for the <<_masters>> above.
```
Comment (Techassi) [inline docs/modules/hbase/pages/usage-guide/operations/graceful-shutdown.adoc:20]:
```suggestion
They use the same mechanism described above.
```

File: docs/modules/hbase/pages/usage-guide/operations/graceful-shutdown.adoc [hunk 3]
```diff
@@ -1,7 +1,41 @@
...
+They will also acknowledge the graceful shutdown with a message in the logs:
```
Comment (Techassi) [inline docs/modules/hbase/pages/usage-guide/operations/graceful-shutdown.adoc:21]:
```suggestion
In contrast to the Master servers, they will, however, acknowledge the graceful shutdown with a message in the logs:
```

File: docs/modules/hbase/pages/usage-guide/operations/graceful-shutdown.adoc [hunk 4]
```diff
@@ -1,7 +1,41 @@
...
+2023-10-11 12:38:05,059 INFO  [shutdown-hook-0] regionserver.ShutdownHook: Shutdown hook starting; hbase.shutdown.hook=true; fsShutdownHook=org.apache.hadoop.fs.FileSystem$Cache$ClientFinalizer@5875de6a
+2023-10-11 12:38:05,060 INFO  [shutdown-hook-0] regionserver.HRegionServer: ***** STOPPING region server 'test-hbase-regionserver-default-0.test-hbase-regionserver-default.kuttl-test-topical-parakeet.svc.cluster.local,16020,1697027870348' *****
+----
+
+== RestServers
+
+As a default, RestServers have `5 minutes` to terminate gracefully.
+
+They go through the same mechanism as documented for the <<_masters>> above.
+They will also acknowledge the graceful shutdown with a message in the logs:
```
Comment (Techassi) [inline docs/modules/hbase/pages/usage-guide/operations/graceful-shutdown.adoc:34]: Maybe use the same phrasing as mentioned above.

File: rust/crd/src/lib.rs [hunk 5]
```diff
@@ -306,6 +318,9 @@ pub struct HbaseConfig {
     pub logging: Logging<Container>,
     #[fragment_attrs(serde(default))]
     pub affinity: StackableAffinity,
+    /// Time period Pods have to gracefully shut down, e.g. `30m`, `1h` or `2d`. Consult the operator documentation for details.
```
Comment (Techassi) [inline rust/crd/src/lib.rs:321]: Add a newline between fields.

File: rust/operator-binary/src/hbase_controller.rs [hunk 6]
```diff
@@ -219,6 +219,10 @@ pub enum Error {
     FailedToCreatePdb {
         source: crate::operations::pdb::Error,
     },
+    #[snafu(display("failed to configure graceful shutdown"))]
```
Comment (Techassi) [inline rust/operator-binary/src/hbase_controller.rs:222]: Add a newline between error variants.

**Review by Techassi — CHANGES_REQUESTED** [review]
Some minor nits.

File: docs/modules/hbase/pages/usage-guide/operations/graceful-shutdown.adoc [hunk 7]
```diff
@@ -1,7 +1,41 @@
...
 
-Graceful shutdown of HBase nodes is either not supported by the product itself
-or we have not implemented it yet.
+You can configure the graceful shutdown as described in xref:concepts:operations/graceful_shutdown.adoc[].
 
-Outstanding implementation work for the graceful shutdowns of all products where this functionality is relevant is tracked in
-https://github.com/stackabletech/issues/issues/357
+== Masters
+
+As a default, masters have `20 minutes` to terminate gracefully.
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply docs/modules/hbase/pages/usage-guide/operations/graceful-shutdown.adoc]: +1 for shutdown

**Comment by sbernauer** [comment]
Please have a look at https://github.com/stackabletech/hbase-operator/commit/5a274008ce793c0b2e3cd2b5ff84941cbb291655 and https://github.com/stackabletech/hdfs-operator/commit/153b2c205a0c330967110a28c2b1c168cb275290

File: rust/crd/src/lib.rs [hunk 8]
```diff
@@ -233,13 +239,19 @@ impl HbaseRole {
                 storage: HbaseStorageConfigFragment {},
             },
         };
+        let graceful_shutdown_timeout = match &self {
```
Comment (Techassi) [inline rust/crd/src/lib.rs:242]: Add newline.

**Review by Techassi — CHANGES_REQUESTED** [review]
Final nit :)

File: rust/crd/src/lib.rs (hunk 8 above)
Comment (sbernauer) [reply rust/crd/src/lib.rs]: done

**Comment by sbernauer** [comment]
https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hbase-operator-it-custom/107/

**Review by Techassi — APPROVED** [review]

**Review by siegfriedweber — DISMISSED** [review]
LGTM

**Comment by sbernauer** [comment]
Final test run: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hbase-operator-it-custom/110/

**Review by Techassi — APPROVED** [review]

**Comment by sbernauer** [comment]
Started https://ci.stackable.tech/job/hbase-operator-it-custom/111/

**Comment by sbernauer** [comment]
Tests passed :rocket:

**Pull request merged**